obsreport/obsreport_exporter.go

```diff
@@ -0,0 +1,216 @@
+	ctx context.Context,
+	span *trace.Span,
+	numExportedSpans int,
+	numDroppedSpans int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForTraceExporter(
+			ctx, numExportedSpans, numDroppedSpans)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedSpans,
+		err,
+		traceData,
+	)
+}
+
+// StartMetricsExportOp is called at the start of an Export operation.
+// The returned context should be used in other calls to the obsreport functions
+// dealing with the same export operation.
+func StartMetricsExportOp(
+	ctx context.Context,
+	exporter string,
+) (context.Context, *trace.Span) {
+	return traceExportDataOp(
+		exporterContext(ctx, exporter),
+		exporter,
+		exportMetricsOperationSuffix)
+}
+
+// EndMetricsExportOp completes the export operation that was started with
+// StartMetricsExportOp.
+func EndMetricsExportOp(
+	ctx context.Context,
+	span *trace.Span,
+	numExportedPoints int,
+	numExportedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	numDroppedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForMetricsExporter(
+			ctx, numExportedTimeSeries, numDroppedTimeSeries)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedPoints,
+		err,
+		metricsData,
+	)
+}
+
+// exporterContext adds the keys used when recording observability metrics to
+// the given context returning the newly created context. This context should
+// be used in related calls to the obsreport functions so metrics are properly
+// recorded.
+func exporterContext(
+	ctx context.Context,
+	exporter string,
+) context.Context {
+	if useLegacy {
+		ctx = observability.ContextWithExporterName(ctx, exporter)
+	}
+
+	if useNew {
+		ctx, _ = tag.New(ctx, tag.Upsert(
+			tagKeyExporter, exporter, tag.WithTTL(tag.TTLNoPropagation)))
+	}
+
+	return ctx
+}
+
+// traceExportDataOp creates the span used to trace the operation. Returning
+// the updated context and the created span.
+func traceExportDataOp(
+	exporterCtx context.Context,
+	exporterName string,
+	operationSuffix string,
+) (context.Context, *trace.Span) {
+	spanName := exporterPrefix + exporterName + operationSuffix
+	return trace.StartSpan(exporterCtx, spanName)
+}
+
+// endExportOp records the observability signals at the end of an operation.
+func endExportOp(
+	exporterCtx context.Context,
+	span *trace.Span,
+	numExportedItems int,
+	err error,
+	dataType dataType,
+) {
+	numSent := numExportedItems
+	numFailedToSend := 0
+	if err != nil {
+		numSent = 0
+		numFailedToSend = numExportedItems
+	}
+
+	var sentMeasure, failedToSendMeasure *stats.Int64Measure
+	var sentItemsKey, failedToSendItemsKey string
+	switch dataType {
+	case traceData:
+		sentMeasure = mExporterSentSpans
+		failedToSendMeasure = mExporterFailedToSendSpans
+		sentItemsKey = SentSpansKey
+		failedToSendItemsKey = FailedToSendSpansKey
```

Comment:
This appears to be unnecessary work if span.IsRecordingEvents() is false.

Reply:
Correct, but is it worth checking it again below? I will check the effects of this when investigating the other suggested optimizations.

Reply:
I benchmarked this and it is a bit better to do multiple checks, will change accordingly.